### PR TITLE
Added kernel_mode and user_mode features in tracelogging crate

### PR DESCRIPTION
**Pull request by hamzamust**
**Why**
`tracelogging` crate cannot be used in kernel-mode code because of the reliance on Win32 APIs such as `EventRegister`.

**What Changed**
* Added optional `kernel_mode` feature which replaces Win32 APIs with kernel-mode APIs like `EtwRegister`
* Added optional `user_mode` feature for enabling Win32 APIs which is set by default to keep in parity with the existing configuration.
* Made provider static variable public for use across different files in the program.

**How Tested**
Added it to a kernel-mode driver built using [windows-drivers-rs](https://github.com/microsoft/windows-drivers-rs) and tested tracelogging prints on WinDbg.

**Usage**
Enabling `kernel_mode` feature will require the client to link with `NtosKrnl.lib` for definition of `Etw*` APIs.

### CODE REVIEW

File: etw/rust/tracelogging_macros/src/provider_generator.rs [hunk 1]
```diff
@@ -50,6 +50,7 @@ impl ProviderGenerator {
         let prov_tokens = self
             .prov_tree
             // static PROVIDER: ::tracelogging::Provider = unsafe { ... };
+            .add_ident("pub")
```
Comment (Robo210) [inline etw/rust/tracelogging_macros/src/provider_generator.rs:53]: This probably should be in a different PR.

File: etw/rust/tracelogging/Cargo.toml [hunk 2]
```diff
@@ -24,8 +24,10 @@ readme = "README.md"
 rust-version = "1.63"
 
 [features]
-default = ["etw", "macros"]
+default = ["macros", "etw", "user_mode"]
 etw = [] # Logging is enabled if windows && etw.
+user_mode = []
```
Comment (Robo210) [inline etw/rust/tracelogging/Cargo.toml:29]: We've not got a situation where people who explicitly enable the etw and macros features will no longer get events if they pick up this new crate version, unless they also modify their code to enable the user_mode feature.

It would probably be best to instead:
- Add only the kernel_mode feature, and assume if it is not enabled then we are user_mode
- Increment the crate version number

**Review by Robo210 — CHANGES_REQUESTED** [review]

File: etw/rust/tracelogging_macros/src/provider_generator.rs (hunk 1 above)
Comment (idigdoug) [reply etw/rust/tracelogging_macros/src/provider_generator.rs:53]: Making this pub is a bit problematic since the provider must be strictly scoped to the crate and must not be exported from the crate. My idea is that you would declare the provider in the root module so that child modules can use it.

File: etw/rust/tracelogging/Cargo.toml (hunk 2 above)
Comment (idigdoug) [inline etw/rust/tracelogging/Cargo.toml:29]: I'm not sure why we need both user_mode and kernel_mode. We should only need one new flag, `kernel_mode`.

File: etw/rust/tracelogging/src/native.rs [hunk 3]
```diff
@@ -69,10 +69,14 @@ impl ProviderContext {
         {
             result = 50; // ERROR_NOT_SUPPORTED
         }
-        #[cfg(all(windows, feature = "etw"))]
+        #[cfg(all(windows, feature = "etw", feature = "user_mode"))]
```
Comment (idigdoug) [inline etw/rust/tracelogging/src/native.rs:72]: Instead of feature = user_mode, can you use `not(feature = "kernel_mode")`?

File: etw/rust/tracelogging_macros/src/provider_generator.rs (hunk 1 above)
Comment (hamzamust) [reply etw/rust/tracelogging_macros/src/provider_generator.rs:53]: The purpose of making this `pub` was that it was not accessible for different files inside the same crate. 

I was unable to access this static in my `device.rs` after I declared it in my `driver.rs`. The error I received was as follows:

```
device.rs 1 of 1 problem
static `MY_TRACELOGGING_PROVIDER` is private
private static rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics)
driver.rs(9, 1): the static `MY_TRACELOGGING_PROVIDER` is defined here
```

in the code `use driver::MY_TRACELOGGING_PROVIDER;`

File: etw/rust/tracelogging/src/native.rs [hunk 4]
```diff
@@ -69,10 +69,14 @@ impl ProviderContext {
...
         {
             result = unsafe { EventActivityIdControl(_control_code, _activity_id) };
         }
+        #[cfg(all(windows, feature = "etw", feature = "kernel_mode"))]
+        {
+            result = unsafe { EtwActivityIdControl(_control_code, _activity_id) };
+        }
```
Comment (idigdoug) [inline etw/rust/tracelogging/src/native.rs:79]: Please add a new value WindowsKernelMode to the NativeImplementation enumeration, and set the `NATIVE_IMPLEMENTATION` constant accordingly.

File: etw/rust/tracelogging/src/native.rs [hunk 5]
```diff
@@ -272,19 +300,39 @@ impl ProviderContextInner {
...
+        #[cfg(all(windows, feature = "etw", feature = "kernel_mode"))]
+        {
+            let was_busy = self.busy.swap(true, atomic::Ordering::Acquire);
+            if was_busy {
+                result = 0;
+            } else {
+                if self.reg_handle == 0 {
+                    result = 0;
+                } else {
+                    result = unsafe { EtwUnregister(self.reg_handle) };
```
Comment (idigdoug) [inline etw/rust/tracelogging/src/native.rs:329]: Why did this duplicate a large chunk of code instead of just one line?

**Review by idigdoug — CHANGES_REQUESTED** [review]

File: etw/rust/tracelogging_macros/src/provider_generator.rs (hunk 1 above)
Comment (idigdoug) [reply etw/rust/tracelogging_macros/src/provider_generator.rs:53]: Making it pub causes problems. Can you instead put your provider in lib.rs or in some other parent mod so that it doesn't need to be public?

File: etw/rust/tracelogging/src/native.rs (hunk 5 above)
Comment (hamzamust) [reply etw/rust/tracelogging/src/native.rs:324]: Fixed

File: etw/rust/tracelogging/src/native.rs (hunk 4 above)
Comment (hamzamust) [reply etw/rust/tracelogging/src/native.rs:87]: Added WindowsKernelMode enum and set it in NATIVE_IMPLEMENTATION const

File: etw/rust/tracelogging/src/native.rs (hunk 3 above)
Comment (hamzamust) [reply etw/rust/tracelogging/src/native.rs]: Replaced user_mode with not(feature="kernel_mode")

File: etw/rust/tracelogging/Cargo.toml (hunk 2 above)
Comment (hamzamust) [reply etw/rust/tracelogging/Cargo.toml]: Removed user_mode

File: etw/rust/tracelogging_macros/src/provider_generator.rs (hunk 1 above)
Comment (hamzamust) [reply etw/rust/tracelogging_macros/src/provider_generator.rs]: Removed pub ident

File: etw/rust/tracelogging/Cargo.toml (hunk 2 above)
Comment (hamzamust) [reply etw/rust/tracelogging/Cargo.toml]: * Removed user_mode feature
* Incremented crate version number

File: etw/rust/tracelogging_dynamic/Cargo.toml [hunk 6]
```diff
@@ -28,4 +28,4 @@ default = ["etw"]
 etw = ["tracelogging/etw"] # Logging is enabled if windows && etw.
 
 [dependencies]
-tracelogging = { default-features = false, version = "= 1.2.1", path = "../tracelogging" }
+tracelogging = { default-features = false, version = "= 1.2.2", path = "../tracelogging" }
```
Comment (idigdoug) [inline etw/rust/tracelogging_dynamic/Cargo.toml:31]: tracelogging_dynamic version also needs to be updated.

**Review by idigdoug — APPROVED** [review]

**Review by Robo210 — APPROVED** [review]
